Check for bash version in vim-patch.sh

xylix opened the pull request:
Because of licensing reasons apple is not going to upgrade their installed bash version anytime soon. Checking for an unsupported version and informing the user about the need to update their bash is probably the easiest workaround we can do. Could also compromise and make the script be compatible with antique bash, but that doesn't sound necessary.

I'd gladly also add this warning to any other scripts / to a wrapper script if there's other places in the neovim build / contribution chain where a bash version check could be useful. Any pointers about this?

Fixes  #12069

xylix committed:
Check for bash version in vim-patch.sh

Wordof-worf reviewed (approved):
навигации между коммитами в запросе на включение.

justinmk commented:
why not check `$BASH_VERSION` ? if the first char is less than 4 then it's too old

justinmk commented on scripts/vim-patch.sh:
```diff
@@ -5,6 +5,12 @@ set -u
 set -p
 
+# Ensure that the user has a bash that supports -A
+if ! declare -A ; then
```
doesn't this echo a bunch of noise?

xylix commented:
I didn't know how to do it in bash, but figured it out now. Now it looks in the $BASH_VERSION.

blueyed commented on scripts/vim-patch.sh:
```diff
@@ -5,6 +5,12 @@ set -u
 set -p
 
+# Ensure that the user has a bash that supports -A
+if [ "${BASH_VERSION::1}" -lt 4  ]; then
```
Should use `BASH_VERSINFO[0]` ("The major version number (the release).", see `bash(1)`).
Otherwise it fails for bash 10 in some years.
Have not checked when `BASH_VERSINFO` was added though..

xylix replied:
https://github.com/bminor/bash/blob/master/CHANGES#L9370
Versinfo was added in bash 2.0, so 1996. Fixed now.

blueyed commented on scripts/vim-patch.sh:
```diff
@@ -5,6 +5,12 @@ set -u
 set -p
 
+# Ensure that the user has a bash that supports -A
+if [ "${BASH_VERSINFO[0]}" -lt 4  ]; then
```
```suggestion
if [[ "${BASH_VERSINFO[0]}" -lt 4  ]]; then
```

blueyed commented on scripts/vim-patch.sh:
```diff
@@ -5,6 +5,12 @@ set -u
 
+# Ensure that the user has a bash that supports -A
+if [ "${BASH_VERSINFO[0]}" -lt 4  ]; then
+  echo "Update your bash version to one that supports -A syntax (>3)"
```
```suggestion
  echo "This script requires bash version 3 or later (you have ${BASH_VERSION})." >&2
```

blueyed reviewed (commented):
LGTM in general, but I think the message can be improved, especially with regard to writing to stderr.

xylix committed:
Apply suggestions from code review

Improve error message for unsupported bash version, use double square bracket operator

Co-Authored-By: Daniel Hahler <github@thequod.de>

xylix commented:
Applied feedback changes

blueyed reviewed (approved):
Thanks!

justinmk commented on scripts/vim-patch.sh:
```diff
@@ -5,6 +5,12 @@ set -u
 
+# Ensure that the user has a bash that supports -A
+if [[ "${BASH_VERSINFO[0]}" -lt 4  ]]; then
+  echo "This script requires bash version 3 or later (you have ${BASH_VERSION})." >&2
```
should be "requires bash 4+". version 3 is too old
